### PR TITLE
fused_moe: remove incorrect topk adjustment in get_2stage_cfgs for EP

### LINKED ISSUE #5595: [ROCm][AITER]: get_2stage_cfgs: topk -= int(is_ep) incorrectly strips Fused Shared Expert column, producing wrong kernel config and gibberish output
### Problem Description

  **Description**

  When expert_mask is provided and the last column of topk_ids contains a real (non-masked) shared expert ID — as done by vLLM's Fused Shared Expert (FSE) feature (VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1) — the MoE layer produces entirely gibberish output with no errors or warnings.

**Root Cause**

  In get_2stage_cfgs, the EP convention strips one column from the topk count before the config lookup:

  EP convention: callers append one always-masked fake-expert slot to `topk_ids`, so runtime `topk` is routed_topk + 1. Tuned configs are keyed
  on routed_topk; strip the fake slot before building the lookup key.
```
  topk -= int(is_ep)

  is_ep is derived in fused_moe_() as:

  is_ep = expert_mask is not None

```
  The assumption is that when expert_mask is present, the last topk_ids column always holds a fake/masked expert (expert_mask[fake_id] = 0). This is correct for the standard EP path.

  However, vLLM's FSE feature pre-fills that same extra column with a real shared expert global ID (e.g., 256 for GLM with 256 routed experts), where expert_mask[256] = 1 (the shared expert is active on all ranks). The kernel config is therefore looked up for topk=8 effective experts, but the kernel is dispatched with 9 real experts, producing incorrect output.

 Reproducer

  Tested on the container directly — no vLLM required:

```
  import torch, inspect
  import aiter
  from aiter import ActivationType, QuantType
  from aiter.fused_moe import get_2stage_cfgs, fused_moe

  dtypes = aiter.dtypes

  # 1. Confirm fused_moe has no shared_expert_id parameter
  params = list(inspect.signature(fused_moe).parameters.keys())
  print('fused_moe params:', params)
  print('has shared_expert_id:', 'shared_expert_id' in params)

  # 2. Demonstrate topk decrement bug
  # GLM: 256 routed experts, top_k=8, 1 real shared expert -> topk_ids has 9 cols
  topk_actual = 9

  print('\n=== is_ep=True (triggered because expert_mask is not None) ===')
  cfg_ep = get_2stage_cfgs(
      token=128, model_dim=4096, inter_dim=2048,
      expert=256, topk=topk_actual,
      dtype=dtypes.fp8, q_dtype_a=dtypes.fp8, q_dtype_w=dtypes.fp8,
      q_type=QuantType.per_Token.value,
      use_g1u1=True, activation=ActivationType.Silu.value,
      doweight_stage1=False, hidden_pad=0, intermediate_pad=0,
      is_ep=True,
  )
  print('Config (internal topk=8, kernel processes 9 experts):', cfg_ep)

  print('\n=== is_ep=False (correct: shared expert is real, not fake) ===')
  cfg_fse = get_2stage_cfgs(
      token=128, model_dim=4096, inter_dim=2048,
      expert=256, topk=topk_actual,
      dtype=dtypes.fp8, q_dtype_a=dtypes.fp8, q_dtype_w=dtypes.fp8,
      q_type=QuantType.per_Token.value,
      use_g1u1=True, activation=ActivationType.Silu.value,
      doweight_stage1=False, hidden_pad=0, intermediate_pad=0,
      is_ep=False,
  )
  print('Config (topk=9, matches kernel):', cfg_fse)
  print('\nConfigs differ (proves wrong config selected under EP):', cfg_ep != cfg_fse)
```

  **Output:**
  fused_moe params: ['hidden_states', 'w1', 'w2', 'topk_weight', 'topk_ids', 'expert_mask',
    'activation', 'quant_type', 'doweight_stage1', 'w1_scale', 'w2_scale', 'a1_scale',
    'a2_scale', 'block_size_M', 'num_local_tokens', 'moe_sorting_dispatch_policy', 'dtype',
    'hidden_pad', 'intermediate_pad', 'bias1', 'bias2', 'splitk', 'swiglu_limit', 'beta',
    'linear_beta', 'gate_mode']
  has shared_expert_id: False

  === is_ep=True (triggered because expert_mask is not None) ===
  Config (internal topk=8, kernel processes 9 experts): MOEMetadata(... block_m=128 ...)

  === is_ep=False (correct: shared expert is real, not fake) ===
  Config (topk=9, matches kernel): MOEMetadata(... block_m=128 ...)

  Configs differ (proves wrong config selected under EP): True

**Expected Behaviour**

  When the extra topk_ids column contains a real expert (expert_mask[id] = 1), topk must not be decremented. The kernel config should be looked up for topk=9 to match the actual number of experts being processed.

**Suggested Fix**

  Add a has_fused_shared_expert: bool = False parameter to fused_moe() (and fused_moe_()) that bypasses the topk -= int(is_ep) decrement when the extra column is a real shared expert:

  `in get_2stage_cfgs / fused_moe_`:
```
  if is_ep and not has_fused_shared_expert:
      topk -= 1
```
 Alternatively, accept shared_expert_id: int = -1 directly in fused_moe() and handle the shared expert internally without embedding it in topk_ids at all — this is the approach taken in vLLM PR [vllm-project/vllm#53161](https://github.com/vllm-project/vllm/pull/53161).

 **Workaround**

  Disable FSE: VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=0. The shared expert then runs as a separate MLP call, which is correct but loses the fusion benefit.

### Operating System

22.04.5 LTS 

### CPU

Intel(R) Xeon(R) Platinum 8480C

### GPU

AMD Instinct MI300X

### ROCm Version

7.2.3

### Installation Method

Python wheels

### Installed ROCm Packages / Versions

<details>
<summary>Installed ROCm packages / versions</summary>

```
Paste output here
```

</details>

### ROCm Component

_No response_

### Steps to Reproduce

_No response_

### (Optional for Linux users) Output of rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by psakhamo**
## Summary

  Removes `topk -= int(is_ep)` from `get_2stage_cfgs` in `aiter/fused_moe.py`.

  ## Root Cause

  When `VLLM_ROCM_USE_AITER_FUSED_MOE_SHARED_EXPERT=1` is set together with EP (`expert_mask is not None`), the removed line incorrectly reduces
  `topk` by 1 before building the config lookup key. This strips the Fused Shared Expert column from the key, causing a wrong tuned-config to be
  selected (or a cache miss falling back to untuned defaults).

  The adjustment was originally intended to strip the fake always-masked expert slot that EP callers append to `topk_ids`. That convention no longer
   applies uniformly — fused shared expert uses the last expert slot as a real expert, so subtracting 1 is invalid in that path.

  ## Change

  Removes 4 lines from `get_2stage_cfgs` — the comment block and the `topk -= int(is_ep)` mutation.

  ## Test Plan
  - Reproduce with EP + `VLLM_ROCM_USE_AITER_FUSED_MOE_SHARED_EXPERT=1` and confirm correct config is selected
  - Run `op_tests/test_moe_2stage.py` to verify no regression on standard EP path

  Fixes #5595

### CODE REVIEW

**Comment by psakhamo** [comment]
Validation on GLM-5.3-FP8 DP8+EP8 — MI300X (gfx942) — vLLM v0.28.0

Node: 8× MI300X, gfx942 | vLLM: v0.28.0+rocm723 | ROCm: 7.2.3
Model: zai-org/GLM-5.3-FP8 (744B total / 39B active, 256 routed + 1 shared experts)
Serving mode: DP8+EP8 (--data-parallel-size 8 --enable-expert-parallel)
Additional patches in image: PR#4365 (qh64 page_size=64 gate) + PR#5563 (__half JIT fix)

**Launch Command**

```
docker run -d --name glm53-fse-test \
    --network host --device /dev/kfd --device /dev/dri \
    --group-add video --group-add render \
    --ipc host --shm-size 128g \
    --cap-add SYS_PTRACE --cap-add IPC_LOCK \
    --security-opt seccomp=unconfined \
    --ulimit nofile=1048576:1048576 \
    -v /lvm-data/psakhamo/glm-5.3-fp8:/mnt/models/GLM-5.3-FP8:ro \
    -v /lvm-data/psakhamo/aiter_jit_glm53_fse:/opt/vllm_cache \
    -v /lvm-data/psakhamo/aiter_home_glm53_fse:/root/.aiter \
    -e HF_HUB_OFFLINE=1 -e TRANSFORMERS_OFFLINE=1 \
    -e VLLM_GCN_ARCH=gfx942 \
    -e VLLM_ROCM_USE_AITER=1 \
    -e VLLM_ROCM_USE_AITER_MOE=1 \
    -e VLLM_ROCM_USE_AITER_MLA=1 \
    -e VLLM_ROCM_USE_AITER_FP8BMM=0 \
    -e VLLM_ROCM_USE_AITER_FP4BMM=0 \
    -e VLLM_USE_V2_MODEL_RUNNER=1 \
    -e VLLM_ROCM_USE_AITER_FUSED_MOE_SHARED_EXPERT=1 \
    -e AITER_BATON_TIMEOUT=7200 \
    -e VLLM_ENGINE_READY_TIMEOUT_S=21600 \
    -e VLLM_HANDSHAKE_TIMEOUT_MINS=30 \
    -e HSA_NO_SCRATCH_RECLAIM=1 \
    -e AITER_JIT_DIR=/opt/vllm_cache/aiter_jit \
    -e VLLM_CACHE_ROOT=/opt/vllm_cache/vllm \
    vllm-glm53-v028-wideep-fse:latest \
    vllm serve /mnt/models/GLM-5.3-FP8 \
        --served-model-name glm-5-3-fp8 \
        --tensor-parallel-size 1 \
        --data-parallel-size 8 --data-parallel-size-local 8 \
        --enable-expert-parallel --trust-remote-code \
        --quantization fp8 --kv-cache-dtype fp8 \
        --max-model-len 202752 --max-num-batched-tokens 8192 \
        --no-enable-prefix-caching \
        --gpu-memory-utilization 0.9 \
        --distributed-timeout-seconds 7200 \
        --enable-auto-tool-choice --tool-call-parser glm47 \
        --port 30000
```
**Reasoning Test**

```
curl -s http://localhost:30000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{"model":"glm-5-3-fp8","messages":[{"role":"user","content":"What is 15 * 47? Think step by step."}],"max_tokens":512,"temperature":0}' \
  | python3 -c "
import sys,json; d=json.load(sys.stdin); msg=d['choices'][0]['message']
print(msg.get('content','')); print('finish_reason:', d['choices'][0]['finish_reason'])"
```

**Output:**

```
# Calculating 15 × 47

**Step 1: Break down 47 into easier parts**
47 = 40 + 7

**Step 2: Multiply each part by 15**
- 15 × 40 = 600
- 15 × 7 = 105

**Step 3: Add the results together**
600 + 105 = **705**

**Verification** (using a different breakdown):
- 15 × 47 = (10 + 5) × 47
- 10 × 47 = 470
- 5 × 47 = 235
- 470 + 235 = 705 ✓

**Answer: 15 × 47 = 705**

finish_reason: stop
```

✅ Correct answer | ✅ Clean finish_reason: stop | ✅ All 8 DP workers completed warmup without GPU faults | ✅ VLLM_ROCM_USE_AITER_FUSED_MOE_SHARED_EXPERT=1 accepted without errors

Without PR#5734, the topk -= int(is_ep) line would silently corrupt the config lookup key when the fused shared expert path is active on EP, causing wrong tuned-config selection. With the patch, config selection is correct on gfx942 + DP8+EP8.

**Comment by yadaish** [comment]
LGTM @valarLip topk -= int(is_ep) should be applied to fake expert mode. Remove it .

**Review by valarLip — APPROVED** [review]
